### PR TITLE
AMD v9 pull 2

### PR DESCRIPTION
**Pull request by babumoger**
Support AMD processors in cmt-cat tool

## Description
AMD is working on to support RDT feature(AMD calls it QoS feature) in next generation hardware. Right now the tool intel-cmt-cat does not support AMD. Majority of these features are similar in both the vendors. However, there are some differences in the way some of these features are implemented in each vendor. We needed to separate those functions and define the vendor specific versions of these functions. This pull request is a first attempt in that direction.

The specification for this AMD feature is available at
https://developer.amd.com/wp-content/resources/56375.pdf

## Affected parts

- [X] library
- [X] pqos utility
- [X] rdtset utility
- [ ] other: (please specify)

## Motivation and Context

Idea here is to have a one tool to support both Intel and AMD. It becomes easy to maintain one tool for both the vendors. It also benefits the OS vendors.

## How Has This Been Tested?

This pull request only has partial set of changes. These patches does not add AMD support yet. So, I have tested on Intel box to make sure it does not introduce any regressions.
Hardware configuration of the Intel system.

lscpu

Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Byte Order: Little Endian
Address sizes: 46 bits physical, 48 bits virtual
CPU(s): 32
On-line CPU(s) list: 0-31
Thread(s) per core: 2
Core(s) per socket: 16
Socket(s): 1
NUMA node(s): 1
Vendor ID: GenuineIntel
CPU family: 6
Model: 85
Model name: Intel(R) Xeon(R) Gold 5218 CPU @ 2.30GHz
Stepping: 7
CPU MHz: 1001.145
CPU max MHz: 3900.0000
CPU min MHz: 1000.0000
BogoMIPS: 4600.00
Virtualization: VT-x
L1d cache: 32K
L1i cache: 32K
L2 cache: 1024K
L3 cache: 22528K
NUMA node0 CPU(s): 0-31
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge
mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall
nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl
xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl
vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2
x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm
abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin
ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept
vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm
cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt
avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc
cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku
ospke avx512_vnni md_clear flush_l1d arch_capabilities

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
User interface for the tool has not changed. So old documentation still holds good. But we may need to update the doc once the AMD support is fully integrated.

### CODE REVIEW

File: lib/allocation.c [hunk 1]
```diff
@@ -1446,7 +1446,7 @@ hw_alloc_reset(const enum pqos_cdp_config l3_cdp_cfg,
 		for (j = 0; j < l3cat_id_num; j++) {
                         unsigned core = 0;
 
-			ret = pqos_cpu_get_one_core(m_cpu, l3cat_ids[j], &core);
+			ret = pqos_cpu_get_one_by_l3cat_id(m_cpu, l3cat_ids[j], &core);
```
Comment (aleksinx) [inline lib/allocation.c]: Travis reported the following code style issue:

allocation.c:1449: WARNING: line over 80 characters
#1449: FILE: allocation.c:1449:
+			ret = pqos_cpu_get_one_by_l3cat_id(m_cpu, l3cat_ids[j], &core);

File: lib/os_allocation.c [hunk 2]
```diff
@@ -1055,8 +1055,18 @@ verify_mba_id(const unsigned mba_id, const struct pqos_cpuinfo *cpu)
...
+/*
+ * @brief Set the bitmask for a given COS
+ *
+ * @param [in] l3cat_id L3 cat identifier
+ * @param [in] num_cos Total number of COS
+ * @param [in] ca Bitmask information to set
+ *
+ * @return Operations status
+ * @retval PQOS_RETVAL_OK on success
+ */
```
Comment (aleksinx) [inline lib/os_allocation.c]: Documentation should be moved to os_allocation.h header file

File: lib/allocation.c [hunk 3]
```diff
@@ -1485,14 +1486,21 @@ hw_alloc_reset(const enum pqos_cdp_config l3_cdp_cfg,
...
                 /**
 		 * Go through all L3 CAT ids and reset MBA class definitions
                  * 0 is the default MBA COS value in linear mode.
                  */
-		for (j = 0; j < l3cat_id_num; j++) {
+		for (j = 0; j < mba_id_num; j++) {
                         unsigned core = 0;
 
-			ret = pqos_cpu_get_one_core(m_cpu, l3cat_ids[j], &core);
+			ret = pqos_cpu_get_one_core(m_cpu, mba_ids[j], &core);
```
Comment (aleksinx) [inline lib/allocation.c]: It should be pqos_cpu_get_one_core_by_mba_id

File: lib/os_allocation.c [hunk 4]
```diff
@@ -1490,7 +1490,7 @@ os_l2ca_get_min_cbm_bits(unsigned *min_cbm_bits)
 }
 
 int
-os_mba_set(const unsigned socket,
+os_mba_set(const unsigned mba_id,
```
Comment (aleksinx) [inline lib/os_allocation.c]: Please update header file

File: lib/os_allocation.c [hunk 5]
```diff
@@ -1615,7 +1615,7 @@ os_mba_set(const unsigned socket,
 }
 
 int
-os_mba_get(const unsigned socket,
+os_mba_get(const unsigned mba_id,
```
Comment (aleksinx) [inline lib/os_allocation.c]: Please update header file

**Comment by babumoger** [comment]
> Please update documentation of the flowing public library API
> pqos_l3ca_set
> pqos_l3ca_get
> pqos_mba_set
> pqos_mba_get



Please update documentation of the flowing public library API
pqos_l3ca_set
pqos_l3ca_get
pqos_mba_set
pqos_mba_get

Please update documentation of the flowing public library API
pqos_l3ca_set
pqos_l3ca_get
pqos_mba_set
pqos_mba_get

File: lib/os_allocation.c (hunk 4 above)
Comment (babumoger) [reply lib/os_allocation.c]: yes. again.. These are taken care in further patches.

File: lib/os_allocation.c (hunk 5 above)
Comment (babumoger) [reply lib/os_allocation.c]: yes. again.. These are taken care in further patches. I will have to add further patches if need to fix it.

File: lib/os_allocation.c (hunk 2 above)
Comment (babumoger) [reply lib/os_allocation.c]: ok. will do.

**Comment by babumoger** [comment]
> > Please update documentation of the flowing public library API
> > pqos_l3ca_set
> > pqos_l3ca_get
> > pqos_mba_set
> > pqos_mba_get

Yes. I have updated these documentation in subsequent patches.  I will have to pull those patches i I need to fix these.  Marcel wanted to keep the pull request smaller.

File: lib/os_allocation.c (hunk 5 above)
Comment (aleksinx) [reply lib/os_allocation.c]: This is very small change in os_allocation.h file header that can be included in this pull request

**Comment by aleksinx** [comment]
Ok, the public API can be updated in the next pull request.

Please, update documentation of internal API because implementation probably won't need be changed in subsequent patches.

The last thing that I would line to include in this pull request is pqos_cpu_get_one_core_by_mba_id API

File: lib/os_allocation.c (hunk 4 above)
Comment (babumoger) [reply lib/os_allocation.c]: pulled couple more patches to address it. thanks.

File: lib/os_allocation.c (hunk 5 above)
Comment (babumoger) [reply lib/os_allocation.c]: ok.  Took care of it.

**Comment by aleksinx** [comment]
Please keep in mind that Python library wrapper should also be updated with the new API. This could be done in separate pull request.

**Comment by babumoger** [comment]
Ok. Will look into it.
Thanks

**Comment by aleksinx** [comment]
We misunderstood and the last patch is not needed. The rest of changes have been merged into master.

**Comment by babumoger** [comment]
Ok. I will remove the last patch. Working on next pull request now.

**Comment by babumoger** [comment]
The commits for this pull request has been merged.

**Comment by babumoger** [comment]
Michal,
  I have sent a new pull request with remaining set of changes. Please
take a look. I have looked at Python library wrapper. It appeared to build
ok. Not sure what needs to be changed there.

**Comment by aleksinx** [comment]
The new API should be added to the Python wrapper.

**Pull request closed without merging**